cpp/frontend_pyreadline.cc

```diff
@@ -4,73 +4,235 @@
 
 #include "_build/detected-cpp-config.h"
 
+#if HAVE_READLINE
+  #include <readline/history.h>
+  #include <readline/readline.h>
+#endif
+
 namespace py_readline {
 
+static Readline* gReadline = nullptr;
+
+// Assuming readline 4.0+
+#if HAVE_READLINE
+
+  #define completion_matches(x, y) \
+    rl_completion_matches((x), ((rl_compentry_func_t*)(y)))
+
+static char* do_complete(const char* text, int state) {
+  if (gReadline->completer_ == nullptr) {
+    return nullptr;
+  }
+  rl_attempted_completion_over = 1;
+  Str* gc_text = StrFromC(text);
+  Str* result = completion::ExecuteReadlineCallback(gReadline->completer_,
+                                                    gc_text, state);
+  if (result == nullptr) {
+    return nullptr;
+  }
+
+  // According to https://web.mit.edu/gnu/doc/html/rlman_2.html#SEC37, readline
+  // will free any memory we return to it.
+  return strdup(result->data());
+}
+
+static char** completion_handler(const char* text, int start, int end) {
+  rl_completion_suppress_append = 0;
+  gReadline->begidx_ = start;
+  gReadline->endidx_ = end;
+  return completion_matches(text, *do_complete);
+}
+
+static void display_matches_hook(char** matches, int num_matches,
+                                 int max_length) {
+  if (gReadline->display_ == nullptr) {
+    return;
+  }
+  auto* gc_matches = Alloc<List<Str*>>();
+  // It isn't clear from the readline documentation, but matches[0] is the
+  // completion text and the matches returned by any callbacks start at index 1.
+  for (int i = 1; i <= num_matches; i++) {
```

Comment:
This was the reason for that weird indentation error you saw in this screencast:
[![asciicast](https://asciinema.org/a/557003.svg)](https://asciinema.org/a/557003)

You can see the correct behavior in this new one:
[![asciicast](https://asciinema.org/a/557335.svg)](https://asciinema.org/a/557335)

Reply:
Awesome!  I'm excited to try this out!